src/react-components/auth/AuthContext.js

```diff
@@ -0,0 +1,96 @@
+import React, { createContext, useState, useEffect, useCallback } from "react";
+import PropTypes from "prop-types";
+import configs from "../../utils/configs";
+
+// TODO: We really shouldn't include these dependencies on every page. A dynamic import would work better.
+import jwtDecode from "jwt-decode";
+import AuthChannel from "../../utils/auth-channel";
+import { connectToReticulum } from "../../utils/phoenix-utils";
+
+export const AuthContext = createContext();
+
+export function AuthContextProvider({ children, store }) {
+  const signIn = useCallback(
+    async email => {
+      const authChannel = new AuthChannel(store);
+      const socket = await connectToReticulum();
+      authChannel.setSocket(socket);
+      const { authComplete } = await authChannel.startAuthentication(email);
+      await authComplete;
+
+      // TODO: Doing all of this just to determine if the user is an admin seems unnecessary. The auth callback should have the isAdmin flag.
+      const retPhxChannel = socket.channel("ret", { hub_id: "index", token: store.state.credentials.token });
+
+      const perms = await new Promise(resolve => {
+        retPhxChannel.join().receive("ok", () => {
+          retPhxChannel.push("refresh_perms_token").receive("ok", ({ perms_token }) => {
+            const perms = jwtDecode(perms_token);
+            retPhxChannel.leave();
+            resolve(perms);
+          });
+        });
+      });
+
+      configs.setIsAdmin(perms.postgrest_role === "ret_admin");
+      store.update({ credentials: { isAdmin: perms.postgrest_role === "ret_admin" } });
+    },
+    [store]
+  );
+
+  const verify = useCallback(
+    async authParams => {
+      const authChannel = new AuthChannel(store);
+      const socket = await connectToReticulum();
+      authChannel.setSocket(socket);
+      await authChannel.verifyAuthentication(authParams.topic, authParams.token, authParams.payload);
+    },
+    [store]
+  );
+
+  const signOut = useCallback(
+    async () => {
+      store.update({ credentials: { token: null, email: null, isAdmin: false } });
+      await store.resetToRandomDefaultAvatar();
+    },
+    [store]
+  );
+
+  const [context, setContext] = useState({
+    isSignedIn: !!store.state.credentials.token,
```

Comment:
I think we talked about this before but since these are all just properties of `store` maybe they would just be part of a `useHubsStore` hook. Though I can see an argument for breaking up the different parts of store like credentials, preferences, and tutorial flow.

Reply:
Yeah, I think we can rethink this in the API PR. I already have changed it a bit.